misk/src/main/kotlin/misk/digester/MergingDigest.kt

```diff
@@ -0,0 +1,351 @@
+package misk.digester
+
+import misk.proto.tdigest.Centroid
+import misk.proto.tdigest.MergingDigestData
+import kotlin.math.PI
+import kotlin.math.asin
+
+class MergingDigest {
+
+    //amount of compression to data
+    var compression: Double = 0.0
+
+    //main list of centroids
+    var mainCentroids: MutableList<Centroid> = mutableListOf()
+    //total weight of unmerged main centroids
+    var mainWeight: Double = 0.0
+
+    //centroids that have been added but not yet merged into main list
+    var tempCentroids: MutableList<Centroid> = mutableListOf()
+    //total weight of unmerged temp centroids
+    var tempWeight: Double = 0.0
+
+    var min: Double = Double.POSITIVE_INFINITY
+    var max: Double = Double.NEGATIVE_INFINITY
+
+    constructor()
+
+    constructor(compression: Double) {
+      this.compression = compression
+    }
+
+    constructor (
+      compression: Double,
+      mainCentroids: MutableList<Centroid>,
+      min: Double,
+      max: Double
+    ) {
+      this.compression = compression
+      this.mainCentroids = mainCentroids
+      this.min = min
+      this.max = max
+    }
+
+  /**
+   * Initializes a new merging t-digest using the given compression parameter.
+   * Lower compression values result in reduced memory consumption and less
+   * precision, especially at the median. Values from 20 to 1000 are recommended
+   * in Dunning's paper.
+   */
+  fun newMerging(compressionValue: Double): MergingDigest {
+    // this is a provable upper bound on the size of the centroid list
+
+    //Todo: Restrict size of mainCentroid - is this even necessary? Its properties not seem to be used in stripe's merging_digest
+   // var sizeBound = ((PI * compressionValue / 2) + 0.5).toInt()
+
+    return MergingDigest(compressionValue)
+  }
+
+  /**
+   *  NewMergingFromData returns a MergingDigest with values initialized from
+   * MergingDigestData.  This should be the way to generate a MergingDigest
+   * from a serialized protobuf.
+   */
+  fun newMergingFromData(mergingDigestData: MergingDigestData): MergingDigest {
+
+    var td = MergingDigest(
+        mergingDigestData.compression,
+        mergingDigestData.main_centroids.toMutableList(),
+        mergingDigestData.min,
+        mergingDigestData.max)
+
+    td.mainWeight = 0.0
+    for (c in td.mainCentroids) {
+      td.mainWeight += c.weight
+    }
+
+    return td
+  }
+
+  /**
+   * function taken from stripes implementation of merging_digest
+   * https://github.com/stripe/veneur/blob/master/tdigest/merging_digest.go
+   */
+  fun estimateTempBuffer(compressionValue: Double): Int {
```

Comment:
`private`

Only the upper-cased methods from the Go implementation (like `Add`) should be ported as public.